### DISCUSSION TITLE
Help Needed: C# Scripting from .dll

### DISCUSSION THREAD

**Discussion by rem-bou** [TabularEditor/TabularEditor3#867] [Other Topics]
Hi Team,

I have an issue when trying to create a custom dll to call reusable code.
I want to be able to create methods where I can add measure to folders, create hierarchies, format columns etc...
For that I have used some help from bernatagulloesbrina on his [blog posts](https://www.esbrina-ba.com/how-to-write-a-c-script-for-tabular-editor-part-2-in-script-classes/) and other ressources on internet as well as this issue section repository here.

When I use the class in Tabular editor (with some updating to catter for the model and to be able the class, etc...). It works. But when I am using the dll as reference to use the cade I get an error. Error happens when trying to add a second column to a hierrachy I have just created.

The dll is quite simple:

```
  using System;
  using System.Collections.Generic;
  using System.Linq;
  using System.Text;
  using System.Threading.Tasks;
  
  using TabularEditor.TOMWrapper;
  using TabularEditor.Scripting;
  using static TabularEditor.Scripting.ScriptHelper;
  using System.Reflection;
  
  
  namespace TE_Helper
  {
      public class TE_Wrapper
      { //public static
          public Model Model; // Define the model that is going to be used.
          TabularEditor.UI.UITreeSelection Selected; // To be able to use SelectTable() or SelectColumn(), etc...
  
          public void AddColumnToHierarch(string tableName, string HierarchyName, string ColumnName)
          {
              // to be used by calling function: AddColumnToHierarch(tableName, HierarchyName, ColumnName);
              // Check if hierarchy exists
              if (!Model.AllHierarchies.Any(h => h.Name == HierarchyName))
              {
                  var h = Model.Tables[tableName].AddHierarchy(HierarchyName);
              }
              // Check if column exists and if not already in hierarchy
              if (Model.Tables[tableName].Columns.Any(t => t.Name == ColumnName) && !Model.Tables[tableName].Hierarchies[HierarchyName].GetChildren().Any(t => t.Name == ColumnName))
              {
                  Model.Tables[tableName].Hierarchies[HierarchyName].AddLevel(ColumnName);
              }
  
          }
      }
  }
```

and the code I use to call the function in Tabular Editor is as followed:

```
#r "C:\Users\xxxxxxxxxx\AppData\Local\TabularEditor3\TE_Helper.dll"
using TE_Helper;

var tableName = "Date";

var MyHelpersScripts = new TE_Helper.TE_Wrapper();
MyHelpersScripts.Model = Model;

// Creation Hierarchy
var HierarchyName = "__Calendar Year Hierarchy";
MyHelpersScripts.AddColumnToHierarch(tableName, HierarchyName, "Calendar Year");
MyHelpersScripts.AddColumnToHierarch(tableName, HierarchyName, "Calendar Quarter");
```

and here is the error message.
![image](https://user-images.githubusercontent.com/102403055/223659424-1864e96d-dd42-46f7-995b-3b8ee57353d0.png)

would you be able to help me please? I assume I am not referencing properly tabular editor dll but I can't figure it out myself.

Thank you

**Comment by mlonsk** [comment]
Hi @rem-bou 
Have you seen this blogpost by Bernat Agulló where he runs through how to build customer DLLs: https://www.esbrina-ba.com/bringing-tabular-editor-c-scripts-to-the-next-level/

I think he has since then shelved this way of working due to it not being convenient for distribution: https://www.esbrina-ba.com/writing-tabular-editor-c-scripts-like-a-boss/

Hope this can help you figure out the issue, otherwise reach out again and we can have another look

**Reply by rem-bou** [reply]
Hi @mlonsk 

I have used Bernat Agulló's blog post to help me with getting started indeed. I am so glad he made these posts!

After further digging, it looks like the issue is on my last if statement:
`!Model.Tables[tableName].Hierarchies[HierarchyName].GetChildren().Any(t => t.Name == ColumnName)`
Paticularly the usage of Any or Where method to filter/check with using dll from TE2.

I have checked the code above using TE3 dll and it works as expected. If using C# script windows in TE2 or TE3, code works as well with no issues.If I understand correctly, we can't use dll from TE3 due to licencing (fair enough!) - Is it correct?

As a "fix", I am using a try/catch but I really don't like this way. Would you have any suggestion instead?
